meeko/polymer.py

```diff
@@ -2478,14 +2599,53 @@ class ResidueTemplate:
     """
 
     def __init__(self, smiles, link_labels=None, atom_names=None):
+
+        # Initializer attributes 
+        self.link_labels = link_labels
+        self.atom_names = atom_names
+
+        # (JSON-bound) computed attributes
         ps = Chem.SmilesParserParams()
         ps.removeHs = False
         mol = Chem.MolFromSmiles(smiles, ps)
         self.check(mol, link_labels, atom_names)
         self.mol = mol
-        self.link_labels = link_labels
-        self.atom_names = atom_names
-        return
+    
+    # region JSON-interchange functions
+    @classmethod
+    def json_encoder(cls, obj: "ResidueTemplate") -> Optional[dict[str, Any]]:
+        output_dict = {
+            "mol": rdMolInterchange.MolToJSON(obj.mol),
+            "link_labels": obj.link_labels,
+            "atom_name": obj.atom_names,
+        }
+        return output_dict
+    
+    # Keys to check for deserialized JSON 
+    expected_json_keys = {"mol", "link_labels", "atom_name"}
+    
+    @classmethod
+    def _decode_object(cls, obj: dict[str, Any]): 
+
+        # Converting ResidueTemplate init values that need conversion
+        deserialized_mol = rdkit_mol_from_json(obj.get("mol"))
+        # do not write canonical smiles to preserve original atom order
+        if deserialized_mol: 
+            mol_smiles = rdkit.Chem.MolToSmiles(deserialized_mol, canonical=False)
+        # if dry json (data) is supplied
+        else:
+            mol_smiles = obj.get("smiles")
+
+        link_labels = convert_to_int_keyed_dict(obj.get("link_labels"))
+
+        # Construct a ResidueTemplate object
+        residue_template = cls(mol_smiles, None, obj.get("atom_name"))
```

Comment:
Renaming `atom_names` to `atom_name` makes it fail to read polymer JSON written by meeko v0.6.1. I thing this was to reuse the `from_dict` function from the base class to lead the templates, which is great, but we must be able to read polymer JSON from v0.6.1. I'll push a (failing) test for this.

I'm also concerned about the `obj.get(key)` vs `obj[key]`. I think it's better to have a missing key error, which is easy to debug, than some trouble with an expected None with a more complex error traceback.

Reply:
`obj.get()` here are being forgiving for (manually prepared) templates that are missing properties. Could we allow missing keys ("link_labels" and/or "atom_name") in template files? 
I will review check of keys and usage of `obj.get()` in other classes.

Reply:
The keys are checked in `from_json`. The missing or excess keys will be logged there. I forgive some keys (maybe not all, that's bad - I should have made these more visible) in the decoding logic to allow lenient parsing. In other words, the error handling is in 'from_json`. The class-specific decoding function `_decode_object` is generally forgiving

Reply:
Do you prefer an early return (no attempt to create obj at all) for missing keys? Currently the mismatched keys are logged if non-critical